app/controllers/favorites_controller.rb

```diff
@@ -3,36 +3,23 @@ class FavoritesController < ApplicationController
   before_action :login_required
 
   def index
+    @page_title = 'Favorites'
+
     return if params[:view] == 'bucket'
-    unless current_user.favorites.present?
+    unless (@favorites = current_user.favorites).exists?
       @posts = []
       return
     end
 
-    @posts = Post.unscoped
-    arel = Post.arel_table
-    where_calc = nil
-
-    current_user.favorites.each do |favorite_rec|
-      new_calc = nil
-      if favorite_rec.favorite_type == User.to_s
-        post_ids = PostAuthor.where(user_id: favorite_rec.favorite_id).where(joined: true).pluck(:post_id)
-        new_calc = arel[:id].in(post_ids)
-      elsif favorite_rec.favorite_type == Post.to_s
-        new_calc = arel[:id].eq(favorite_rec.favorite_id)
-      elsif favorite_rec.favorite_type == Board.to_s
-        new_calc = arel[:board_id].eq(favorite_rec.favorite_id)
-      end
-      if where_calc.nil?
-        where_calc = new_calc
-      else
-        where_calc = where_calc.or(new_calc)
-      end
-    end
+    user_favorites = @favorites.where(favorite_type: User).select(:favorite_id)
+    author_posts = PostAuthor.where(user_id: user_favorites, joined: true).select(:post_id)
+    board_favorites = @favorites.where(favorite_type: Board).select(:favorite_id)
+    post_favorites = @favorites.where(favorite_type: Post).select(:favorite_id)
 
+    @posts = Post.unscoped
+    @posts = @posts.where(id: author_posts).or(@posts.where(id: post_favorites)).or(@posts.where(board_id: board_favorites))
     @posts = posts_from_relation(@posts.where(where_calc).ordered)
```

Comment:
Should this be:

```rb
@posts = Posts.where(id: author_posts).or(@posts.where(id: post_favorites)).or(@posts.where(board_id: board_favorites))
@posts = posts_from_relation(@posts.ordered)
```

? It seems to be failing in CI.

I'm also not sure if `favorite_type: User` will work directly, or if you'll need to put it in quotes – seems plausible that'll be necessary.

This looks neat, though!